#!/bin/bash


#This script will take the .kraken and .fasta files indicated by the user to create a .csv table that contains the 
#number of contig (NODE), its length, its coverage, its GC %, and the taxonomic lineage generated by the 
#taxonomic assignment done with kraken 2 (.kraken file)

#The output of the script is a csv table named after the .kraken file

#To run the script it is necesary that this shell script and the python script "gc_and_taxa.py" are in the same directory

#Auxiliary file "taba1.csv" contains the NODE (contig), its length, and its coverage (obtained from the .kraken file)
#Auxiliary file "aux1.txt" contains the taxonomic ids (obtained from the .kraken file)
#Auxiliary file "aux2.txt" contains the unique taxonomic ids (obtained from the .kraken file)


file_kraken=$1
file_contigs=$2

echo "NODE,LENGTH,COVERAGE,GC,SUPERKINGDOM,PHYLUM,ORDER,FAMILY,GENUS,SPECIES,SUB-SPECIES" > ${file_kraken/%.kraken/.csv}
cut -f2 ${file_kraken} | cut -d'_' -f 2,4,6 | sed -e 's/_/,/g' > taba1.csv
cut -f3 ${file_kraken} > aux1.txt
cut -f3 ${file_kraken} | sort | uniq | awk '{if($1>3)print $1}' > aux2.txt

python gc_and_taxa.py $2 >> ${file_kraken/%.kraken/.csv}

rm aux1.txt
rm aux2.txt

rm taba1.csv

